internal/app/app.go

```diff
@@ -127,25 +108,20 @@ func (s *Server) encodeURL(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		genLink = "http://" + r.Host + "/" + genLink
-
-		s.db.Mutex.Lock()
-		_, ok := s.db.pairs[genLink]
-		s.db.Mutex.Unlock()
-
+		err = s.db.SaveURL(context.Background(), link, genLink)
```

Comment:
```suggestion
		err = s.db.SaveURL(r.Context(), link, genLink)
```
Контекст вычислений в рамках обработки запроса пользователя необходимо брать из запроса. Если пользователь в какой-то момент разорвал соединение, контекст протухнет и потащит за собой остановку всех выполняющихся в рамках запроса действий.